sdk/src/androidTest/java/com/mapbox/maps/MapIntegrationTest.kt

```diff
@@ -88,6 +92,322 @@ class MapIntegrationTest {
+            .build()
+        )
+        mapboxMap.loadStyleUri(Style.MAPBOX_STREETS) { style ->
+          // prepare layer
+          val layer = circleLayer("layer", "source") {
+            circleColor("red")
+            circleRadius(10.0)
+          }
+          // async loading
+          val source = geoJsonSource("source") {
+            geometry(Point.fromLngLat(0.0, 0.0))
+          }
+          // sync method must clear async queue and not allow current running async task apply changes
+          source.data(
+            """
+            {
+              "type": "FeatureCollection",
+              "features": [
+                {
+                  "type": "Feature",
+                  "geometry": {
+                    "type": "Point",
+                    "coordinates": [
+                      0.1,
+                      0.1
+                    ]
+                  },
+                  "properties": {
+                    "icon-image": "cafe-15",
+                    "is-draggable": true
+                  }
+                }
+              ]
+            }
+            """.trimIndent()
+          )
+          style.addSource(source)
+          style.addLayer(layer)
+          mapView.postDelayed(
+            {
+              mapboxMap.queryRenderedFeatures(
+                ScreenCoordinate(mapView.width / 2.0, mapView.height / 2.0),
+                RenderedQueryOptions(listOf("layer"), null)
+              ) { result ->
+                if (result.value?.size == 1) {
+                  countDownLatch.countDown()
+                }
+              }
+            },
+            1_000L
+          )
+        }
+        mapView.onStart()
+      }
+    }
+    if (!countDownLatch.await(5, TimeUnit.SECONDS)) {
+      throw TimeoutException()
+    }
+  }
+
+  @Test
+  fun testApplySyncDataToGeoJson3() {
+    countDownLatch = CountDownLatch(1)
+    rule.scenario.onActivity {
+      it.runOnUiThread {
+        mapView = MapView(it)
+        mapboxMap = mapView.getMapboxMap()
+        it.frameLayout.addView(mapView)
+        mapboxMap.setCamera(
+          CameraOptions.Builder()
+            .center(Point.fromLngLat(0.2, 0.2))
+            .zoom(9.0)
+            .build()
+        )
+        mapboxMap.loadStyleUri(Style.MAPBOX_STREETS) { style ->
+          // prepare layer
+          val layer = circleLayer("layer", "source") {
+            circleColor("red")
+            circleRadius(10.0)
+          }
+          // async loading
+          val source = geoJsonSource("source") {
+            geometry(Point.fromLngLat(0.0, 0.0))
+          }
+          // sync method must clear async queue and not allow current running async task apply changes
+          source.geometry(Point.fromLngLat(0.1, 0.1))
+          // follow up with new async that must take effect
+          source.geometry(Point.fromLngLat(0.2, 0.2)) {}
+          style.addSource(source)
+          style.addLayer(layer)
+          mapView.postDelayed(
```

Comment:
this might result in flaky execution as it's not a guarantee that anything will be rendered, let's try it and if an issue revisit.

Reply:
It's not for rendering but for async parse to complete 😄 
I hope parsing one point should fit in 1 second otherwise we're in trouble!